core/api/src/main/java/org/trellisldp/api/BinaryService.java

```diff
@@ -36,6 +43,30 @@
      */
     CompletableFuture<Void> setContent(BinaryMetadata metadata, InputStream stream);
 
+    /**
+     * Set the content for a binary object using a digest algorithm.
+     *
+     * @implSpec The default implementation will compute a digest while processing the {@code InputStream}.
+     *           A {@code null} response value indicates that the digest algorithm is not supported. In the
+     *           case of a non-supported algorithm, the content will not be sent along to the binary service.
+     * @param metadata the binary metadata
+     * @param stream the context
+     * @param algorithm the digest algorithm
+     * @return the new completion stage containing the server-computed digest. A {@code null} value
+     *         can be understood to mean that the digest algorithm is unsupported.
+     */
+    default CompletableFuture<String> setContent(final BinaryMetadata metadata, final InputStream stream,
```

Comment:
I like the basic signature here, but are you sure we wouldn't rather this be a `byte[]` rather than a `String`? I ask because `String` implies an encoding, but we don't really have that here. I think that's why `MessageDigest::digest`, etc. return a `byte[]` instead.

Reply:
Using a `byte[]` would be fine. The HTTP layer has a Base64-encoded `String` (which comes from the incoming `Digest` header), but we can just do that conversion in the HTTP layer.

Reply:
Yeah, that's exactly what I meant.

Reply:
Thinking about this some more, we should change the signature of the `CompletableFuture<String> calculateDigest(IRI identifier, String algorithm)` to also return `byte[]`

Reply:
Oh, duh. Yep!